File: backpack/p416/Solution.java
class Solution {
    public boolean canPartition(int[] nums) {
        if(nums == null || nums.length < 2)
            return false;
        
        int total = 0;
        for(int x : nums)
            total += x;
        
        if(total % 2 != 0)
            return false;
        
        int target = total / 2;
        int[][] DP = new int[nums.length + 1][target + 1];
        DP[0][0] = 1;
        
        for(int i = 1; i < nums.length + 1; i++)
            DP[i][0] = 1;
        
        for(int i = 1; i < nums.length + 1; i++) {
            int num = nums[i-1];
            for (int j = 1; j < target + 1; j++) {
                if(DP[i-1][j] == 1 || (j >= num && DP[i-1][j-num] == 1)) {
                    DP[i][j] = 1;
                }
            }
        }
        
        return DP[nums.length][target] == 1;
    }
}

File: backpack/p416/Solution.py
class Solution:
    def canPartition(self, nums):
        """
        :type nums: List[int]
        :rtype: bool
        """
        if len(nums) < 2:
            return False

        total = sum(nums)
        if total % 2 != 0:
            return False

        target = total // 2
        size = len(nums)
        DP = [[False for _ in range(target + 1)]
              for _ in range(size + 1)]
        DP[0][0] = True

        for i in range(1, size + 1):
            DP[i][0] = True

        for j in range(1, target + 1):
            DP[0][j] = False

        for i in range(1, size + 1):
            num = nums[i-1]
            for j in range(1, target + 1):
                DP[i][j] = DP[i-1][j] or DP[i-1][j-num]

        return DP[size][target]
